drivers/usb/device/usb_dc_native_posix_adapt.c

```diff
@@ -0,0 +1,455 @@
+
+	if (recv(connfd, busid, 32, 0) != sizeof(busid)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return errno;
+	}
+
+	if (send(connfd, &header, sizeof(header), 0) != sizeof(header)) {
+		SYS_LOG_ERR("send() header failed: %s", strerror(errno));
+		return errno;
+	}
+
+	send_device(desc, connfd);
+
+	return 0;
+}
+
+extern struct usb_desc_header __usb_descriptor_start[];
+
+void usbip_start(void)
+{
+	struct sockaddr_in srv;
+	unsigned char attached;
+	int listenfd, connfd;
+	const u8_t *desc;
+	int reuse = 1;
+
+	SYS_LOG_DBG("Starting");
+
+	/*
+	 * Do not use usb_get_device_descriptor();
+	 * to prevent double string fixing
+	 */
+	desc = (const u8_t *)__usb_descriptor_start;
+	if (!desc) {
+		SYS_LOG_ERR("Descriptors are not set");
+		posix_exit(EXIT_FAILURE);
+	}
+
+	listenfd = socket(PF_INET, SOCK_STREAM, 0);
+	if (listenfd < 0) {
+		SYS_LOG_ERR("socket() failed: %s", strerror(errno));
+		posix_exit(EXIT_FAILURE);
+	}
+
+	if (setsockopt(listenfd, SOL_SOCKET, SO_REUSEADDR,
+		       (const char *)&reuse, sizeof(reuse)) < 0) {
+		SYS_LOG_WRN("setsockopt() failed: %s", strerror(errno));
+	}
+
+	memset(&srv, 0, sizeof(srv));
+	srv.sin_family = AF_INET;
+	srv.sin_addr.s_addr = htonl(INADDR_ANY);
+	srv.sin_port = htons(USBIP_PORT);
+
+	if (bind(listenfd, (struct sockaddr *)&srv, sizeof(srv)) < 0) {
+		SYS_LOG_ERR("bind() failed: %s", strerror(errno));
+		posix_exit(EXIT_FAILURE);
+	}
+
+	if (listen(listenfd, SOMAXCONN) < 0) {
+		SYS_LOG_ERR("listen() failed: %s", strerror(errno));
+		posix_exit(EXIT_FAILURE);
+	}
+
+	while (true) {
+		struct sockaddr_in client_addr;
+		socklen_t client_addr_len = sizeof(client_addr);
+
+		connfd = accept(listenfd, (struct sockaddr *)&client_addr,
+				&client_addr_len);
+		if (connfd < 0) {
+			SYS_LOG_ERR("accept() failed: %s", strerror(errno));
+			posix_exit(EXIT_FAILURE);
+		}
+
+		connfd_global = connfd;
+
+		SYS_LOG_DBG("Connection: %s", inet_ntoa(client_addr.sin_addr));
+
+		/* Set attached 0 */
+		attached = 0;
+
+		while (true) {
+			struct usbip_header cmd;
+			struct usbip_header_common *hdr = &cmd.common;
+			int read;
+
+			if (!attached) {
+				struct op_common req;
+
+				read = recv(connfd, &req, sizeof(req), 0);
+				if (read != sizeof(req)) {
+					SYS_LOG_WRN("small packet, %u", read);
+					break;
+				}
+
+				hexdump((u8_t *)&req, sizeof(req));
+
+				SYS_LOG_DBG("Code: 0x%x", ntohs(req.code));
+
+				switch (ntohs(req.code)) {
+				case OP_REQ_DEVLIST:
+					handle_device_list(desc, connfd);
+					break;
+				case OP_REQ_IMPORT:
+					if (!handle_import(desc, connfd)) {
+						attached = 1;
+					}
+					break;
+				default:
+					SYS_LOG_ERR("Unhandled code: 0x%x",
+						    ntohs(req.code));
+					break;
+				}
+
+				continue;
+			}
+
+			/* Handle attached case */
+
+			read = recv(connfd, hdr, sizeof(*hdr), 0);
+			if (read != sizeof(*hdr)) {
+				SYS_LOG_ERR("recv failed: %d", read);
+				break;
+			}
+
+			hexdump((u8_t *)hdr, sizeof(*hdr));
```

Comment:
Please move this line above L:399, so that hexdump will also be printed in the error case.